shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroup.java

```diff
@@ -0,0 +1,84 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine;
+
+import android.content.Context;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import io.flutter.embedding.engine.dart.DartExecutor.DartEntrypoint;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class is experimental. Please do not ship production code using it.
+ *
+ * <p>Represents a collection of {@link io.flutter.embedding.engine.FlutterEngine}s who share
+ * resources among each other to allow them to be created faster and with less memory than calling
+ * the {@link io.flutter.embedding.engine.FlutterEngine}'s constructor multiple times.
+ *
+ * <p>When creating or recreating the first {@link io.flutter.embedding.engine.FlutterEngine} in the
+ * FlutterEngineGroup, the behavior is the same as creating a {@link
+ * io.flutter.embedding.engine.FlutterEngine} via its constructor. When subsequent {@link
+ * io.flutter.embedding.engine.FlutterEngine}s are created, resources from an existing living {@link
+ * io.flutter.embedding.engine.FlutterEngine} is re-used.
+ *
+ * <p>Deleting a FlutterEngineGroup doesn't invalidate its existing {@link
+ * io.flutter.embedding.engine.FlutterEngine}s, but it eliminates the possibility to create more
+ * {@link io.flutter.embedding.engine.FlutterEngine}s in that group.
+ */
+public class FlutterEngineGroup {
+
+  public FlutterEngineGroup(@NonNull Context context) {
+    this.context = context;
+  }
+
+  private final Context context;
+  /* package */ @VisibleForTesting final List<FlutterEngine> activeEngines = new ArrayList<>();
```

Comment:
Should these be WeakPtr?  Is onEngineDestroy going to be called if we have a reference here?  The unit tests are calling destroy directly, I'm not sure if that mimics actual usage.

Reply:
Ummm, let me think a bit about whether I can weave it into the FlutterEngineCache mechanism a bit.

Reply:
Thinking about this a bit more:

a- this has to be orthogonal to the FlutterEngineCache on Android. Not all the engine in the cache must come from the same group and we can't assume we can spawn from any engines in the cache. The cache's purpose explicitly has a lifecycle mandate and this does not.
b- though the last point made me think for here and on iOS about the name. I don't think this class will ever have a lifecycle mandate, which makes it not really a group. Should this just be called FlutterEngineGroupFactory? Since building new ones will be the only thing this ever does?

Reply:
Oops, I didn't answer your original question. Generally, people always call release/destroy type methods manually and not use finalizers since it's indeterministic. So none of our embeddings destroy resources as a hook to the finalizer. Therefore things with no references left to it are not really "gone", especially with statics and JNI and circular references etc. So a GC initiated reference removal is almost impossible because things almost always leak without an explicit destruction. And this is already hooked to explicit destructions.